lib/connectors/memory.js

```diff
@@ -679,6 +681,52 @@ Memory.prototype.updateAttributes = function updateAttributes(model, id, data, o
   }
 };
 
+Memory.prototype.replace = function replace(model, id, data, options, cb) {
+  var self = this;
+  if (!id) {
+    var err = new Error('You must provide an id when replacing!');
+    if (cb) {
+      return cb(err);
+    } else {
+      throw err;
+    }
+  }
+
+  // Do not modify the data object passed in arguments
+  // data = Object.create(data);
+
+  this.setIdValue(model, data, id);
+  var cachedModels = this.collection(model);
+  var modelData = cachedModels && this.collection(model)[id];
+  if (modelData) {
+    var id = this.getIdValue(model, data);
+    this.collection(model)[id] = serialize(data);
+    this.saveToFile(data, function (err) {
+      cb(err, self.fromDb(model, data));
+    });
+  } else {
+    cb(new Error('Could not replace. Object with id ' + id + ' does not exist!'));
+  }
+};
+
+Memory.prototype.replaceOrCreate = function (model, data, options, callback) {
+  var self = this;
+  this.exists(model, self.getIdValue(model, data), options, function (err, exists) {
+    if (exists) {
+      var id = self.getIdValue(model, data);
+      self.collection(model)[id] = serialize(data);
+      self.saveToFile(data, function (err) {
+        callback(err, data, {isNewInstance: false});
+      });
+    } else {
+      self.create(model, data, options, function (err, id) {
+        self.setIdValue(model, data, id);
+        callback(err, data, { isNewInstance: true });
+      });
+    }
+  });
+};
```

Comment:
This `replaceOrCreate` block of code is prone to race condition here and it is not atomic; to solve this I need to have my other PR (strongloop/loopback-datasource-juggler/pull/815) merged to master branch and then merge it to my current branch as well; since, in that patch I implemented synchronous `_findAllSkippingIncludes` which can be used  instead of `exists` and synchronous `_create` can be used inseatd of `create`...